spark/src/main/scala/org/apache/spark/sql/delta/stats/DataSkippingReader.scala

```diff
@@ -583,6 +588,191 @@ trait DataSkippingReaderBase
+          nullCol <- getStatsColumnOpt(nullCountStatsCol)
+        } yield {
+          val resolvedAttribute = ResolvedPartitionLikeReference(
+            Seq(minStatsCol, maxStatsCol, nullCountStatsCol),
+            minCol.expr,
+            maxCol.expr,
+            nullCol.expr)
+          (minCol.expr, Set(resolvedAttribute))
+        }
+      // For other attribute references, we can't safely rewrite the expression.
+      case SkippingEligibleColumn(_, _) => None
+      // Don't attempt data skipping on a nondeterministic expression, since the value returned
+      // might be different when executed twice on the same input.
+      // For example, rand() > 0.5 would return ~25% of records if used in data skipping, while the
+      // user would expect ~50% of records to be returned.
+      case other if !other.deterministic => None
+      // Inline subquery results to support InSet. The subquery should generally have already been
+      // evaluated.
+      case in: InSubqueryExec =>
+        // Values may not be defined if the subquery has been skipped - we can't apply this filter.
+        in.values().flatMap { possiblyNullValues =>
+          // Rewrite the children of InSubqueryExec, then replace the subquery with an InSet
+          // containing the materialized values.
+          rewriteDataFiltersAsPartitionLikeInternal(in.child, clusteringColumnPaths).flatMap {
+            case (rewrittenChildren, referencedStats) =>
+              Some(InSet(rewrittenChildren, possiblyNullValues.toSet), referencedStats)
+          }
+        }
+      // Don't allow rewriting UDFs - even if deterministic, UDFs might have some unexpected
+      // side effects when executed twice.
+      case _: UserDefinedExpression => None
+      // Pushdown NOT through OR - we prefer AND to OR because AND can tolerate one branch not being
+      // rewriteable.
+      case Not(Or(e1, e2)) =>
+        rewriteDataFiltersAsPartitionLikeInternal(And(Not(e1), Not(e2)), clusteringColumnPaths)
+      // For AND expressions, we can tolerate one side not being eligible for partition-like
+      // data skipping - simply remove the ineligible side.
+      case And(left, right) =>
+        val leftResult = rewriteDataFiltersAsPartitionLikeInternal(left, clusteringColumnPaths)
+        val rightResult = rewriteDataFiltersAsPartitionLikeInternal(right, clusteringColumnPaths)
+        (leftResult, rightResult) match {
+          case (Some((newLeft, statsLeft)), Some((newRight, statsRight))) =>
+            Some((And(newLeft, newRight), statsLeft ++ statsRight))
+          case _ => leftResult.orElse(rightResult)
+        }
+      // For any other expression, recursively rewrite the expression's children unless the
+      // expression might be very expensive to evaluate (JSON parsing, regex, etc.).
+      case other if !other.containsAnyPattern(INVOKE, JSON_TO_STRUCT, LIKE_FAMLIY,
+        REGEXP_EXTRACT_FAMILY, REGEXP_REPLACE) =>
```

Comment:
Should we add the various UDF types as well? They can be arbitrarily expensive, no?
(the other check only rejects a UDF node that we actually reach, not the parents)

Reply:
But actually... wouldn't this approach pessimize predicates that have an AND somewhere deeper, when not all legs of the AND have expensive/bad operations?

Reply:
Yeah, we would prefer to only reject nodes that we reach. Rewrote this to explicitly match against these expression families so that we don't prematurely reject these expressions.